tools/mage/end-to-end.go

```diff
@@ -0,0 +1,154 @@
+// Copyright © 2020 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ttnmage
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/magefile/mage/mg"
+	"github.com/magefile/mage/sh"
+	"github.com/magefile/mage/target"
+)
+
+// EndToEnd namespace
+type EndToEnd mg.Namespace
+
+var (
+	testDatabaseName = "ttn_lorawan_test"
+	databaseURI      = "postgresql://root@localhost:26257/" + testDatabaseName + "?sslmode=disable"
+)
+
+func getTestURL() string {
+	nodeEnv := os.Getenv("NODE_ENV")
+	port := "1885"
+	if nodeEnv == "development" {
+		port = "8080"
+	}
+	return "http://localhost:" + port
+}
+
+func (EndToEnd) runCypress(command string, args ...string) error {
+	return runYarnCommand("cypress", append([]string{command, fmt.Sprintf("--config-file=%s", filepath.Join("config", "cypress.json")), "--config", fmt.Sprintf("baseUrl=%s", getTestURL())}, args...)...)
+}
+
+func (EndToEnd) prepareDB() error {
+	isCI := os.Getenv("CI") == "true"
+	dumpExists := pathExists(filepath.Join(".cache", "sqldump.sql"))
+	ok, err := target.Dir(
+		filepath.Join(".cache", "/sqldump.sql"),
+		filepath.Join("pkg", "identityserver", "store"),
+	)
+	if err != nil {
+		return targetError(err)
+	}
+	if !ok || (isCI && dumpExists) {
+		mg.SerialDeps(EndToEnd.DBRestore)
+		return nil
+	}
+	os.Setenv("TTN_LW_IS_DATABASE_URI", databaseURI)
+	mg.SerialDeps(Dev.DBErase, Dev.DBStart, Dev.InitStack, EndToEnd.DBDump)
+	return nil
+}
+
+func (EndToEnd) prepareBuild() {
+	mg.SerialDeps(Js.Deps, Js.Build)
```

Comment:
`js.Build` should and depends on `js.Deps`

Reply:
I added this because otherwise the `Js.Deps` target would run multiple times in parallel, leading to crashes, due to other dependencies of `Js.Build` depending on `Js.Deps` as well.

Reply:
`mage` should take care of running each dependency at most once - if that does not happen, it's a bug and it should be fixed instead of introducing hacky workarounds.